mcstatus/server.py

```diff
@@ -51,14 +66,24 @@ def port(self) -> int:
         return self.address.port
 
     @classmethod
+    @abstractmethod
     def lookup(cls, address: str, timeout: float = 3) -> Self:
-        """Mimics minecraft's server address field.
+        """Lookup a minecraft server. This can have additional logic in children classes."""
 
-        :param str address: The address of the Minecraft server, like `example.com:19132`
-        :param float timeout: The timeout in seconds before failing to connect.
-        """
-        addr = Address.parse_address(address, default_port=None)
-        return cls(addr.host, addr.port, timeout=timeout)
```

Comment:
We can probably keep this base parsing logic without anything additional as-is, since it's valid for both java and bedorck, except with java, we extend it by an SRV lookup. `Address.parse_address` is simply a function to convert a string address into an `Address` object.

Reply:
No, because `MCServer` is just an abstract class. At now, there is no one who will use this as realization class. This class need only for implement interface.

Also one of reasons why you can't use it as realization class:
```py
>>> MCServer()
TypeError: Can't instantiate abstract class MCServer with abstract methods async_lookup, async_status, lookup, status
```
And of this reason we should raise `NotImplementError` in abstract methods.

Reply:
That doesn't mean abstract classes can't have any concrete implementations, even just a single abstract method and thousands of concrete methods still make that class an abstract class that's not possible to be initialized.

The reason we should include this logic is simply because all child classes use this same logic, and there's no reason to implement that logic in those child classes, if it's already shared. JavaServer only extends this logic by adding an SRV lookup in there, but other than that, it's the very same logic, meaning it does follow the inheritence order.

Raising `NotImplementedError` in the abstract methods is fine, but again, abstract classes can contain concrete methods, not just abstract ones, and if this method does have implementation which is the same in both, there's no reason not to include it in here. For the methods which truly can't be implemented and must remain abstract (such as status), raising `NotImplementedError` is a good idea though.

Reply:
You can't move realization to abstract class because `Java` and `Bedrock` servers have different `default_port` parameter when calling `Address.parse_address`.

Reply:
Simply omit the port, `parse_address` supports port being `None`. This is also how it was done originally from #223. There's no reason for this PR to be changing this behavior anyway

Reply:
If you doesn't specify port in `address` string, will be something like this:
```py
>>> Address.parse_address("127.0.0.1")
ValueError: Given address '127.0.0.1' doesn't contain port and default_port wasn't specified, can't parse.
```
So default port is required and can't be specified in redefed `__init__` method. Of course I can change `Address.parse_address` that, so it will return None instead of ValueError.

Reply:
That's exactly what I would expect from a base `MCServer.lookup` function though. Since this function contains the functionality which is common across both java and bedrock, and both of those would be performing address parsing like this, however we can't know the default port from here, and so for a base mcserver, that should give us an error.

In `BedrockServer` then, this should be overridden and the default port should be set from there, and in `JavaServer`, the same applies, except there should also be the additional SRV lookup.

It's true that since this required both child classes to override this method anyway, it may not even be worth it to provide this base functionality, but it's not an issue to do so, and that was already the way it worked before this PR. Since this PR doesn't have anything to do with the lookup logic, it shouldn't be changing it here.

If you think making the lookup method fully abstract is a better way to go about this, it can be done in another PR, but probably not here, and it's not a huge issue if that's not done, in fact some people may benefit from a simple shared lookup without any default ports.

Reply:
Undo this change in 46689e006cfa0bb1e4731d180097c8b9c1561a50.